Fabric Text components use their text value as the accessibility name by default

chrisglein opened the pull request:
## Description
Text components were showing in the accessibility tree but not with any value. This change has `UIA_NamePropertyId` default to the text value but can be overriden by `accessibilityLabel`. 

### Type of Change
- New feature (non-breaking change which adds functionality)

### Why
Text components were showing in the accessibility tree but not with any value.

Resolves #12030

### What
Added a virtual `DefaultAccessibleName()` that enables individual Fabric components to provide a fallback if `accessibilityLabel` isn't specified. Then `ParagraphComponentView` overrides this to provide its string value.

## Screenshots
Add any relevant screen captures here from before or after your changes. 
| Before | After |
|--------|--------|
| ![image](https://github.com/microsoft/react-native-windows/assets/26607885/4f95ce31-d0e4-40fe-aa4d-b2e1bb0f7178)| ![image](https://github.com/microsoft/react-native-windows/assets/26607885/55d59c10-0767-4eb4-adf0-18f6e82a1325)|

## Testing
- Tested first in the Artifical Chat app
- Then tested in PlaygroundComposition on the Text test page to ensure that 1. Names started showing up, and 2. `accessibilityLabel` was still respected if specified.

## Changelog
Yes? This is somewhere between bug fix (expected behavior not working) and feature (it wasn't working before).

Fabric Text components use their text value as the accessibility name by default (if `accessibilityLabel` is not specified
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/react-native-windows/pull/12039)

acoates-ms commented on vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:
```diff
@@ -247,7 +247,8 @@ HRESULT __stdcall CompositionDynamicAutomationProvider::GetPropertyValue(PROPERT
     case UIA_NamePropertyId: {
       pRetVal->vt = VT_BSTR;
-      auto wideName = ::Microsoft::Common::Unicode::Utf8ToUtf16(props->accessibilityLabel);
+      auto wideName = ::Microsoft::Common::Unicode::Utf8ToUtf16(
```
This would avoid the extra prop access and Utf8ToUtf16 on empty strings:
`auto wideName = props->accessibilityLabel.empty() ? ::Microsoft::Common::Unicode::Utf8ToUtf16(baseView->DefaultAccessibleName()) : L"";`

jonthysell replied:
@chrisglein You're logic looks fine.

@acoates-ms The logic in your example isn't equivalent as it would never use the `accessibilityLabel` value. If the label has a value, you want to use it, not report an empty string.

acoates-ms replied:
Yeah, sorry... read it wrong.

acoates-ms commented on vnext/Microsoft.ReactNative/Fabric/Composition/CompositionViewComponentView.cpp:
```diff
@@ -1072,7 +1072,10 @@ void CompositionBaseComponentView::updateAccessibilityProps(
+      provider,
+      UIA_NamePropertyId,
+      oldViewProps.accessibilityLabel,
+      newViewProps.accessibilityLabel.empty() ? DefaultAccessibleName() : newViewProps.accessibilityLabel);
```
No need to call newViewProps.accessibilityLabel again if we know its empty.

jonthysell replied:
@chrisglein This looks good.

chiaramooney commented:
Didn't see the PR. Left some comments on the commit itself. https://github.com/chrisglein/react-native-windows/commit/a58cc6d29bce5905b59e9c7c27b81cc0f3ac2c69 

Can you add this PR to this issue https://github.com/microsoft/react-native-windows/issues/11901 in the "Name" row. And update status to In Progress?

Can you also verify that you've tested the FastRefresh case?

chrisglein commented:
> Didn't see the PR. Left some comments on the commit itself. [chrisglein@a58cc6d](https://github.com/chrisglein/react-native-windows/commit/a58cc6d29bce5905b59e9c7c27b81cc0f3ac2c69)
> Can you add this PR to this issue #11901 in the "Name" row. And update status to In Progress?
> Can you also verify that you've tested the FastRefresh case?

- Addressed the comments on the commit with an additional commit.
- I've validated fast refresh on adding/changing/removing `accessibilityLabel` (and `accessibilityRole`, since I touched that line). 
  - Should I modify the [static text](https://github.com/microsoft/react-native-windows/blob/2ed6c533349c85289ea2ccc2bc0cc30864dc0cc1/packages/%40react-native-windows/tester/src/js/examples-win/Text/TextExample.windows.tsx#L860-L869) that's there right now to make it interactive with a Switch to add/remove role? Or is that overkill?
- And I've added a note to the issue.

chiaramooney reviewed (approved):
👍 Looks good to me! The switch to turn on/off setting accessibilityRole is probably not needed for now, but would be a good thing to have once we move into adding functional testing for Fabric. So up to you if you have the time.
